### PR TITLE
OCPBUGS-1257: Have keepalived check for haproxy status for API VIP

### PR DESCRIPTION
**Pull request by mandre**
Keepalived was wrongly checking the status of the load-balanced
kube-apiserver, overstepping over HAProxy's responsibility.
Keepalived is responsible for moving the API VIP to a node where HAProxy
is functioning correctly. As a consequence, it should be checking for
the local HAProxy's readiness.

The previous check would make a request to the local API port, effectively
testing that the local HAProxy was correctly distributing traffic,
however, this check could fail if the selected API server was able to
handle the request, unnecessarily forcing a VIP failover.

This becomes important during recovery from an outage. When none of the
kube-apiservers are healthy this health check will fail continuously,
and the API VIP will move uselessly between masters. However the
situation is much worse when only one of the kube-apiservers is up. In
this case there is a high probability that it is overloaded and at least
rate limiting incoming connections. This may lead us to fail the
keepalived health check and fail the VIP over to the next HAProxy. This
will cause all open kube-apiserver connections to reset, even the
established ones. This increases the load on the kube-apiserver and
increases the probability that the health check will fail again.

This commit changes the keepalived API VIP check to look at the
readiness of HAProxy via it's monitoring interface.

### CODE REVIEW

File: templates/master/00-master/on-prem/files/keepalived-script-both.yaml [hunk 1]
```diff
@@ -3,4 +3,4 @@ path: "/etc/kubernetes/static-pod-resources/keepalived/scripts/chk_ocp_script_bo
 contents:
   inline: |
     #!/bin/bash
-    /usr/bin/curl -o /dev/null -kLfs https://localhost:{{`{{ .LBConfig.LbPort }}`}}/readyz && [ -e /var/run/keepalived/iptables-rule-exists ] || /usr/bin/curl -kLfs https://localhost:{{`{{ .LBConfig.ApiPort }}`}}/readyz
+    /usr/bin/curl -o /dev/null -kLfs https://localhost:9444/haproxy_ready && [ -e /var/run/keepalived/iptables-rule-exists ] || /usr/bin/curl -kLfs https://localhost:{{`{{ .LBConfig.ApiPort }}`}}/readyz
```
Comment (cybertron) [inline templates/master/00-master/on-prem/files/keepalived-script-both.yaml:6]: The concern I have here is that it doesn't actually verify that the api is reachable through haproxy. Just because haproxy is running that doesn't mean everything is working correctly. For example, I'm not sure that on startup we require the haproxy backend list to be populated, so it's possible it could be running with nothing behind it, and if it takes the vip in that state then we'll never recover.

TLDR: the fact that this hits an api endpoint was deliberate.

That said, I'm not immediately coming up with a good solution to the scenario you describe. Maybe some sort of backoff so that if the VIP moves a lot in a short period of time we get less aggressive with our health checks? Or just if we detect instability in the api? That might be easier to implement since we could keep track of how many api requests are failing in the monitor, and if we see more than X% failures we increase the fall value in the health check so the vip is less likely to move due to flakes.

Comment (mandre) [reply templates/master/00-master/on-prem/files/keepalived-script-both.yaml:6]: Thanks for the feedback. I'm gonna have to study exactly how the pivot from bootstrap works to see if we can end up in the deadlock situation you're describing. Stay tuned.

Comment (mandre) [reply templates/master/00-master/on-prem/files/keepalived-script-both.yaml:6]: So, I had a look at the bootstrap sequence and I don't think it's going to be a problem.
The API VIP remains on the bootstrap node until the bootstrap node is torn down, since https://github.com/openshift/machine-config-operator/commit/1537a2234580fea0d4e17bbb5b2e998c09cfca4f.
At that point, we have:
- masters have all booted and started their static pods
- the etcd cluster formed with master nodes is healthy
- the bootstrap control plane is up
- cluster-etcd-operator is healthy

For both haproxy and keepalived unicast, we get the list of backends from baremetal-runtimecfg. This list is populated by querying kube for master nodes, meaning it's not possible to be in a situation where we don't have a backend.

We are sure to have all 3 master nodes registered with the cluster for etcd operator to be healthy (and appear in the list of backends) before the VIP moves to a master node.

Comment (mandre) [reply templates/master/00-master/on-prem/files/keepalived-script-both.yaml:6]: I've just pushed [a new commit](https://github.com/openshift/machine-config-operator/pull/3339/commits/339575f7731834b212109626bddb8d5410d25cd2) that should make sure we never end up in a situation like you describe. HAProxy has a way to cause the healthcheck to fail where there is no backends. See https://www.haproxy.com/documentation/hapee/latest/observability/monitoring/monitor-uri/.

File: templates/master/00-master/on-prem/files/haproxy-haproxy.yaml [hunk 2]
```diff
@@ -23,6 +23,7 @@ contents:
       bind :::9444 v4v6
       mode http
       monitor-uri /haproxy_ready
+      monitor fail if { nbsrv(masters) eq 0 }
```
Comment (mdbooth) [inline templates/master/00-master/on-prem/files/haproxy-haproxy.yaml:26]: This looks like exactly what we want here. Have you managed to test it?

Comment (mdbooth) [reply templates/master/00-master/on-prem/files/haproxy-haproxy.yaml:26]: To be clear, we want this to succeed as long as at least 1 master is configured *regardless of whether any of them are healthy*. It should only fail if there are no masters configured. It would be good to test specifically that it will succeed if there are 3 unhealthy masters.

Comment (cybertron) [reply templates/master/00-master/on-prem/files/haproxy-haproxy.yaml:26]: I'm fine with this, but I'm pretty sure it doesn't do what you want and won't solve the problem you're hitting. My reading of "the expression uses the nbsrv fetch method to check the number of servers that are up in the webservers backend." says that if all three servers are marked down in the backend then this will fail. Which, I would argue, is perfectly reasonable behavior, but doesn't solve the flapping problem when the backends are unstable.

File: templates/master/00-master/on-prem/files/haproxy-haproxy.yaml [hunk 3]
```diff
@@ -16,9 +16,11 @@ contents:
       timeout client       86400s
       timeout server       86400s
       timeout tunnel       86400s
+    {{`{{- if gt (len .LBConfig.Backends) 0 }}`}}
     frontend  main
       bind :::{{`{{ .LBConfig.LbPort }}`}} v4v6
       default_backend masters
+    {{`{{- end }}`}}
```
[the remark this answers is not in the document]
Comment (mandre) [reply templates/master/00-master/on-prem/files/haproxy-haproxy.yaml]: Thanks, done.

**Comment by cybertron** [comment]
I still don't like this but I'm also tired of arguing about it so I'm going to close my eyes and pray we never have a problem between haproxy and the apiservers.
/lgtm

**Comment by mandre** [comment]
> I still don't like this but I'm also tired of arguing about it so I'm going to close my eyes and pray we never have a problem between haproxy and the apiservers. /lgtm

Sorry you feel this way, I wished @mdbooth and myself had managed to convince you that's it's indeed how keepalived should implement the healthchecks. If you want to be conservative, we can implement this for OpenStack at first then change if for all platforms if it proves to be stable.

**Comment by sinnykumari** [comment]
Is vsphere job failure unrelated?
/cc @jcpowermac

**Comment by mandre** [comment]
> Is vsphere job failure unrelated? /cc @jcpowermac

It certainly seems this way to me. I'm not finding anything in the job's output that would point at issues with keepalived
/retest

**Comment by mandre** [comment]
@sinnykumari this should be ready to merge when you get a chance.

**Comment by sinnykumari** [comment]
vsphere test is green, thanks @mandre 
/approve
